### DISCUSSION TITLE
A question about captum in Pyg.

### DISCUSSION THREAD

**Discussion by juanshu30** [pyg-team/pytorch_geometric#4168] [General]
Hi,

I ran the sample code in captum_explainability.py and find three choices for explanation on 'edge', 'node' and 'edge and node'. 

I thought the 'node' one would be explanation on node features(find important features) but the output i got is a vector of size 2708(number of nodes), which means it select important nodes to the current nodes, am i right? I am wondering if I can find node feature importance through captum in pyg. Thanks!

**Comment by rusty1s** [comment]
AFAIK, node explanation returns you a matrix of shape `2708 x 1433` (`num_nodes` times `num_features`), which you can post-process to select a set of most important features, e.g., `ig_attr_node.sum(dim=0).topk(k=10)`. Let me know if this helps you. @RBendias might have some further thoughts.

**Reply by juanshu30** [reply]
> AFAIK, node explanation returns you a matrix of shape `2708 x 1433` (`num_nodes` times `num_features`), which you can post-process to select a set of most important features, e.g., `ig_attr_node.sum(dim=0).topk(k=10)`. Let me know if this helps you. @RBendias might have some further thoughts.

Thanks for your reply. In the sample code:

captum_model = to_captum(model, mask_type='node', output_idx=output_idx)

ig = IntegratedGradients(captum_model)
ig_attr_node = ig.attribute(data.x.unsqueeze(0), target=target,
                            additional_forward_args=(data.edge_index),
                            internal_batch_size=1)

there is a output_index, which means it generates explanation for one node, right? I change the code to(remove output_index and replace target with data.y):

captum_model = to_captum(model, mask_type='node')

ig = IntegratedGradients(captum_model)
ig_attr_node = ig.attribute(data.x.unsqueeze(0), target=data.y,
                            additional_forward_args=(data.edge_index),
                            internal_batch_size=1)

Then it gives me the output with dimension 2708*1433. I am just wondering why if I run the sample code, which focus on one node, it gives me an output vector with dimension 2708 instead of a vector of size 1433.

**Reply by RBendias** [reply]
> there is an output_index, which means it generates explanation for one node, right?

It generates an explanation for the **prediction** of a node specified by "output_idx". However, not only the node itself contributes to the prediction but also other nodes around it. For this reason, the attribution includes all nodes and not just the node itself. Does this help with your question?

**Reply by juanshu30** [reply]
> > there is an output_index, which means it generates explanation for one node, right?
> 
> It generates an explanation for the **prediction** of a node specified by "output_idx". However, not only the node itself contributes to the prediction but also other nodes around it. For this reason, the attribution includes all nodes and not just the node itself. Does this help with your question?

Yeah. Thanks for your explanation.
